panther_lights/src/lights_driver_node.cpp

```diff
@@ -59,45 +61,41 @@ DriverNode::DriverNode(const std::string & node_name, const rclcpp::NodeOptions
   chanel_1_.SetGlobalBrightness(global_brightness);
   chanel_2_.SetGlobalBrightness(global_brightness);
 
+  client_callback_group_ =
+    this->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+
   enable_led_control_client_ = this->create_client<SetBoolSrv>(
-    "hardware/led_control_enable", rmw_qos_profile_services_default);
+    "hardware/led_control_enable", rmw_qos_profile_services_default, client_callback_group_);
 
   set_brightness_server_ = this->create_service<SetLEDBrightnessSrv>(
     "lights/set_brightness", std::bind(&DriverNode::SetBrightnessCB, this, _1, _2));
 
-  diagnostic_updater_.setHardwareID("Bumper Lights");
-  diagnostic_updater_.add("Lights driver status", this, &DriverNode::DiagnoseLights);
-
-  RCLCPP_INFO(this->get_logger(), "Node constructed successfully.");
-}
-
-void DriverNode::InitializeSubscribers(const std::shared_ptr<image_transport::ImageTransport> & it)
-{
-  RCLCPP_INFO(this->get_logger(), "Initializing ImageTransport subscribers.");
-
   chanel_1_ts_ = this->get_clock()->now();
-  chanel_1_sub_ = std::make_shared<image_transport::Subscriber>(
-    it->subscribe("lights/channel_1_frame", 5, [&](const ImageMsg::ConstSharedPtr & msg) {
+  chanel_1_sub_ = this->create_subscription<ImageMsg>(
+    "lights/channel_1_frame", 5, [&](const ImageMsg::UniquePtr & msg) {
       FrameCB(msg, chanel_1_, chanel_1_ts_, "channel_1");
       chanel_1_ts_ = msg->header.stamp;
-    }));
+    });
 
   chanel_2_ts_ = this->get_clock()->now();
-  chanel_2_sub_ = std::make_shared<image_transport::Subscriber>(
-    it->subscribe("lights/channel_2_frame", 5, [&](const ImageMsg::ConstSharedPtr & msg) {
+  chanel_2_sub_ = this->create_subscription<ImageMsg>(
+    "lights/channel_2_frame", 5, [&](const ImageMsg::UniquePtr & msg) {
       FrameCB(msg, chanel_2_, chanel_2_ts_, "channel_2");
       chanel_2_ts_ = msg->header.stamp;
-    }));
+    });
+
+  diagnostic_updater_.setHardwareID("Bumper Lights");
+  diagnostic_updater_.add("Lights driver status", this, &DriverNode::DiagnoseLights);
 
-  RCLCPP_INFO(this->get_logger(), "Initialized ImageTransport successfully.");
+  ToggleLEDControl(true);
+
+  RCLCPP_INFO(this->get_logger(), "Node constructed successfully.");
 }
 
 void DriverNode::OnShutdown()
 {
-  ClearLEDs();
-
-  if (led_control_granted_) {
-    ToggleLEDControl(false);
```

Comment:
If the lights driver shut down but the rest of the system operates as before, the LED stays unnecessarily occupied.

Reply:
Ad. `ToggleLEDControl(false)` - There is `wait for service`, so if a service is already unavailable, after a timeout the exception is thrown. We could add a try-catch block to ensure that each cleanup method is properly executed (possibly in the future someone adds something after `ToggleLEDControl` and we may encounter incomplete execution of a cleanup process).

Ad. `ClearLEDs()` - this cleans the state of an internal class attribute and doesn't need to depend on the fact if control is granted or not. 

Putting things in the code it may look somehow like: 

```cpp
void DriverNode::OnShutdown()
{
  ClearLEDs();

  if (led_control_granted_) {
    try {
      ToggleLEDControl(false);
    } catch (const std::runtime_error & e) {
      RCLCPP_DEBUG(this->get_logger(), "Couldn't revoke LED control during shutdown: %s", e.what());
    }
  }
}
```

Reply:
Ad. ToggleLEDControl(false) - actually quite weird thing happens here. The method does not throw, the program just freezes here I'm not sure what causes this. Sooo, try catch won't help :confused: 
Ad. ClearLEDs() - it sends buffer to SPI, but it makes sense to clear spi buffer on shutdown anyway